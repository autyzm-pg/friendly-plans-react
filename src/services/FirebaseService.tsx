import firebase from 'react-native-firebase';
import { Image } from 'react-native-image-crop-picker';

export class FirebaseService {
  /* Upload image to Firebase Storage and return public URL to download it. */
  static uploadUserImage = async (image: string): Promise<string> => {
    const user = firebase.auth().currentUser!;
    const imageRef = firebase
      .storage()
      .ref('users')
      .child(user.uid);
    const snapshot = await imageRef.putFile(image);
    return snapshot.downloadURL as string;
  };

  static updateUserImage = async (image: string): Promise<void> => {
    const photoURL = await FirebaseService.uploadUserImage(image);
    return firebase.auth().currentUser!.updateProfile({ photoURL });
  };
}
