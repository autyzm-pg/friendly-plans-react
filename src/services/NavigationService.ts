import { NavigationActions, NavigationContainerComponent, NavigationState } from 'react-navigation';

import { DialogProps } from 'models';

interface ScreenProps extends Partial<DialogProps> {
  [property: string]: any;
}

export default class NavigationService {
  navigator?: NavigationContainerComponent;

  setTopLevelNavigator = (navigatorRef: NavigationContainerComponent) => {
    this.navigator = navigatorRef;
  };

  navigate = (routeName: string, params?: ScreenProps) => {
    this.navigator!.dispatch(
      NavigationActions.navigate({
        routeName,
        params,
      }),
    );
  };

  goBack = () => {
    this.navigator!.dispatch(NavigationActions.back());
  };

  getActiveRouteName = (navigationState: NavigationState): string => {
    if (!navigationState) {
      return '';
    }
    const route = navigationState.routes[navigationState.index];
    // dive into nested navigators
    if (route.routes) {
      return this.getActiveRouteName(route as NavigationState);
    }
    return route.routeName;
  };
}
