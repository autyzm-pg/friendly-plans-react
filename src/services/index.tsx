import _AnalyticsService from './AnalyticsService';
import _CrashlyticsService from './CrashlyticsService';
import _NavigationService from './NavigationService';

export const NavigationService = new _NavigationService();
export const AnalyticsService = new _AnalyticsService(NavigationService);
export const CrashlyticsService = new _CrashlyticsService();
