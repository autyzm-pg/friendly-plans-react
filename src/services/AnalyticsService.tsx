import firebase from 'react-native-firebase';
import { NavigationState } from 'react-navigation';

import { User } from 'models';
import NavigationService from './NavigationService';

export default class AnalyticsService {
  navigationService: NavigationService;
  user: User | null;
  screenName?: string;

  constructor(navigationService: NavigationService) {
    firebase.analytics().setAnalyticsCollectionEnabled(!__DEV__);
    this.navigationService = navigationService;
    this.user = null;
  }

  setCurrentScreen = (prevState: NavigationState, currentState: NavigationState) => {
    const currentScreen = this.navigationService.getActiveRouteName(currentState);
    const prevScreen = this.navigationService.getActiveRouteName(prevState);

    if (prevScreen !== currentScreen) {
      firebase.analytics().setCurrentScreen(currentScreen, currentScreen);
      this.screenName = currentScreen;
    }
  };

  setCurrentUser = (user: User | null) => {
    this.user = user;
    firebase.analytics().setUserId(this.user ? this.user!.id.toString() : null);
  };
}
