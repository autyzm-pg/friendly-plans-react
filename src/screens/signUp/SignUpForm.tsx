import { FormikProps } from 'formik';
import React from 'react';
import { StyleSheet, View } from 'react-native';
import ImagePicker, { Image } from 'react-native-image-crop-picker';

import { Button, InputItem } from 'components';
import { i18n } from 'locale';
import { NavigationService } from 'services';
import { palette } from 'styles';

import LoadImageButton from './LoadImageButton';
import { SignUpFormData } from './SignUpFormContainer';

const imageOptions = {
  width: 512,
  height: 512,
  compressImageQuality: 0.8,
  cropping: true,
  cropperCircleOverlay: true,
  writeTempFile: true,
};
interface Props extends FormikProps<SignUpFormData> {
  loading?: boolean;
}

export class SignUpForm extends React.PureComponent<Props> {
  navigateToTermsOfUse = () => {
    NavigationService.navigate('TermsOfUse');
  };
  openPicker = async () => {
    try {
      const image: Image = (await ImagePicker.openPicker(imageOptions)) as Image;
      this.props.setFieldValue('imageUrl', image.path);
    } catch (e) {
      // fail gracefully
    } finally {
      this.setState({ loading: false });
    }
  };

  render() {
    const { handleChange, handleBlur, values, handleSubmit, errors, touched, loading } = this.props;

    return (
      <View style={styles.container}>
        <InputItem
          onChangeText={handleChange('name')}
          onBlur={handleBlur('name')}
          value={values.name}
          error={errors.name}
          touched={touched.name}
          placeholder={i18n.t('common:name')}
          autoCapitalize="none"
          keyboardType="email-address"
          textContentType="name"
          blurOnSubmit
          style={styles.input}
          styleContainer={styles.inputContainer}
        />

        <LoadImageButton imageUrl={values.imageUrl} loading={loading} openPicker={this.openPicker} />

        <InputItem
          onChangeText={handleChange('email')}
          onBlur={handleBlur('email')}
          value={values.email}
          error={errors.email}
          touched={touched.email}
          placeholder={i18n.t('common:email')}
          autoCapitalize="none"
          keyboardType="email-address"
          textContentType="emailAddress"
          blurOnSubmit
          style={styles.input}
          styleContainer={styles.inputContainer}
        />

        <InputItem
          onChangeText={handleChange('password')}
          onBlur={handleBlur('password')}
          value={values.password}
          error={errors.password}
          touched={touched.password}
          placeholder={i18n.t('common:password')}
          secureTextEntry
          textContentType="password"
          blurOnSubmit
          style={styles.input}
          styleContainer={styles.inputContainer}
        />

        <Button
          onPress={handleSubmit}
          title={i18n.t('signUp:signUpButton')}
          containerStyle={styles.signUpbuttonContainer}
          loading={loading}
          buttonStyle={styles.signUpButton}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    width: 272,
    alignSelf: 'center',
  },
  signUpbuttonContainer: {
    marginTop: 8,
    marginBottom: 20,
  },
  inputContainer: { borderWidth: 0, marginBottom: 12 },
  input: {
    backgroundColor: palette.background,
    borderRadius: 12,
    height: 44,
  },
  signUpButton: {
    borderRadius: 12,
    height: 44,
  },
});
