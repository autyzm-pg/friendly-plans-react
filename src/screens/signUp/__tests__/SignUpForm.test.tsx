import { shallow } from 'enzyme';
import React from 'react';

import { NavigationService } from 'services';
import { SignUpForm } from '../SignUpForm';

describe('SignUpForm', () => {
  const props = {
    handleChange: jest.fn(),
    handleBlur: jest.fn(),
    values: {
      email: 'email@domain.com',
      password: 'topsecretpassword',
      termsAccepted: true,
    },
    errors: {
      email: '',
      password: '',
      termsAccepted: '',
    },
    touched: {
      email: true,
      password: true,
      termsAccepted: true,
    },
    handleSubmit: jest.fn(),
    setFieldValue: jest.fn(),
  } as any;

  it('should match snapshot', () => {
    const wrapper = shallow(<SignUpForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should navigate to Terms of Use screen', () => {
    const wrapper = shallow(<SignUpForm {...props} />);
    const instance = wrapper.instance() as SignUpForm;
    const spy = jest.spyOn(NavigationService, 'navigate').mockImplementation(jest.fn());
    instance.navigateToTermsOfUse();
    expect(spy).toBeCalledWith('TermsOfUse');
  });

  it('should set field value on check terms of use', () => {
    const wrapper = shallow(<SignUpForm {...props} />);
    const instance = wrapper.instance() as SignUpForm;
    instance.checkTermsofUse(true);
    expect(props.setFieldValue).toBeCalledWith('termsAccepted', true);
  });
});
