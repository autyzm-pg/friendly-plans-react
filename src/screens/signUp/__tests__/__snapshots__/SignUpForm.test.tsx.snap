// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm should match snapshot 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "width": 272,
    }
  }
>
  <InputItem
    autoCapitalize="none"
    blurOnSubmit={true}
    keyboardType="email-address"
    placeholder="Your Name"
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 12,
        "height": 44,
      }
    }
    styleContainer={
      Object {
        "borderWidth": 0,
        "marginBottom": 12,
      }
    }
    textContentType="name"
  />
  <LoadImageButton
    openPicker={[Function]}
  />
  <InputItem
    autoCapitalize="none"
    blurOnSubmit={true}
    error=""
    keyboardType="email-address"
    placeholder="Email address"
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 12,
        "height": 44,
      }
    }
    styleContainer={
      Object {
        "borderWidth": 0,
        "marginBottom": 12,
      }
    }
    textContentType="emailAddress"
    touched={true}
    value="email@domain.com"
  />
  <InputItem
    blurOnSubmit={true}
    error=""
    placeholder="Password"
    secureTextEntry={true}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 12,
        "height": 44,
      }
    }
    styleContainer={
      Object {
        "borderWidth": 0,
        "marginBottom": 12,
      }
    }
    textContentType="password"
    touched={true}
    value="topsecretpassword"
  />
  <Button
    backgroundColor="#6654EB"
    buttonStyle={
      Object {
        "borderRadius": 12,
        "height": 44,
      }
    }
    containerStyle={
      Object {
        "marginBottom": 20,
        "marginTop": 8,
      }
    }
    onPress={[MockFunction]}
    title="Sign up"
  />
</View>
`;
